src/packet/signature/types.rs

```diff
@@ -219,22 +222,23 @@ impl Signature {
     /// - when backsig is true: verify a "Primary Key Binding Signature (type ID 0x19)"
     fn verify_key_binding_internal(
         &self,
-        signing_key: &impl PublicKeyTrait,
-        key: &impl PublicKeyTrait,
+        signer: &impl PublicKeyTrait,
+        signee: &impl PublicKeyTrait,
         backsig: bool,
     ) -> Result<()> {
         debug!(
-            "verifying key binding: {:#?} - {:#?} - {:#?}",
-            self, signing_key, key
+            "verifying key binding: {:#?} - {:#?} - {:#?} (backsig: {})",
+            self, signer, signee, backsig
         );
 
-        let key_id = signing_key.key_id();
-        if let Some(issuer) = self.issuer() {
-            if &key_id != issuer {
+        // don't check issuer/issuer_fp subpacket match for backsigs. the subpackets are inconsistent in the test suite ("Binding signature subpackets"). (why?)
```

Comment:
Specifically, some of the embedded "back signatures" in https://tests.sequoia-pgp.org/#Binding_signature_subpackets seem to use the primary key's fingerprint as the "issuer fingerprint".

This is odd, because that embedded signature is issued by the signing subkey, which the embedded signature binds to the primary key. However, based on those test cases, I assume that such signatures exist in the wild.

My suggestion is: I would entirely drop the `match_identity` check in `verify_key_binding_internal`. I think it's pragmatic to just always check the cryptographic signature, and not to bother with interpreting the issuer subpackets. Key binding signatures are always "self signatures".
There is exactly one candidate key packet that can have created a key binding signature (including embedded back signatures). Verifying the signature with that one candidate key is straightforward.

(This suggestion applies only to key binding signatures within TPKs, as done in `verify_key_binding_internal`. I would keep the `match_identity` check in the three other locations).

Reply:
(linking to https://github.com/rpgp/rpgp/pull/262#pullrequestreview-1887781505, because that thread discusses the same point)